master/src/main/java/org/apache/celeborn/service/deploy/master/SlotsAllocator.java

```diff
@@ -344,19 +484,26 @@ static List<WorkerInfo> generateRackAwareWorkers(List<WorkerInfo> workers) {
   private static List<Integer> roundRobin(
       Map<WorkerInfo, Tuple2<List<PartitionLocation>, List<PartitionLocation>>> slots,
       List<Integer> partitionIds,
-      List<WorkerInfo> workers,
+      List<WorkerInfo> primaryWorkers,
+      List<WorkerInfo> replicaWorkers,
       Map<WorkerInfo, List<UsableDiskInfo>> slotsRestrictions,
       boolean shouldReplicate,
       boolean shouldRackAware,
-      int availableStorageTypes) {
+      int availableStorageTypes,
+      boolean replicaSameAsPrimary) {
+    if (primaryWorkers.isEmpty() || replicaWorkers.isEmpty()) {
+      return partitionIds;
+    }
     // workerInfo -> (diskIndexForPrimaryAndReplica)
     Map<WorkerInfo, Integer> workerDiskIndex = new HashMap<>();
     List<Integer> partitionIdList = new LinkedList<>(partitionIds);
 
-    final int workerSize = workers.size();
-    final IntUnaryOperator incrementIndex = v -> (v + 1) % workerSize;
-    int primaryIndex = rand.nextInt(workerSize);
-    int replicaIndex = rand.nextInt(workerSize);
+    final int primaryWorkersSize = primaryWorkers.size();
+    final int replicaWorkersSize = replicaWorkers.size();
+    final IntUnaryOperator primaryWorkersIncrementIndex = v -> (v + 1) % primaryWorkersSize;
+    final IntUnaryOperator replicaWorkersIncrementIndex = v -> (v + 1) % replicaWorkersSize;
+    int primaryIndex = rand.nextInt(primaryWorkersSize);
+    int replicaIndex = rand.nextInt(replicaWorkersSize);
```

Comment:
Compute `replicaIndex` here only when `shouldReplicate` is true - else `replicaWorkers` can be empty : resulting in div-by-zero

We need to add a test to catch this case.

Reply:
Fixed. We already have a test case where we pass in false for shouldReplicate